fix: strict mode problem - setting value in control constructor

sastan opened the pull request:
Fixes #591

jsumners reviewed (changes requested):
Thank you for the contribution. At a minimum we will need a unit test to cover this. The test should fail prior to the change and pass with the change in place.

However, I'm hesitant to accept any patches to this code path right now. We are very close to v3 (https://github.com/ldapjs/node-ldapjs/projects/2) being ready. In v3, this code lives in https://github.com/ldapjs/controls. If I'm not mistaken, the issue this PR attempts to fix is not present in the new library.

sastan commented:
Added a test that should fail in non-strict mode. I have added `"use strict"` pragma to the test function but I'm not sure if it would trigger the issue. 

We found this problem after bundling our whole app into a single executable with strict mode.

jsumners commented on test/controls/control.test.js:
```diff
@@ -51,3 +51,41 @@ test('parse no value', function (t) {
 })
+
+test('only value getter in sub-class with strict mode', function (t) {
+  // eslint-disable-next-line quotes
+  "use strict"
```
```suggestion
  'use strict'
```

jsumners commented on test/controls/control.test.js:
```diff
@@ -51,3 +51,41 @@ test('parse no value', function (t) {
+  // eslint-disable-next-line quotes
+  "use strict"
+
+  const assert = require('assert-plus')
```
I'd rather not add any new code using this library. Additionally, we don't need a 100% complete control implementation for the test, just what is necessary to trigger the error.

sastan replied:
sure. no problem

sastan commented:
> > I have added `"use strict"` pragma to the test function but I'm not sure if it would trigger the issue.
> 
> Rollback your change (comment it out) in the library code and run your test. If the test fails, it is a good test.

The test wouldn't trigger the issue. It only happens when `'use strict'` is added to lib/controls/control.js. Because there the assignment is made. If I put it there 41 tests are failing.

This issue only happens when ldapjs is bundle with a global strict mode. 

I understand if you don't want to change existing code. We are using patch-package in our app to apply this fix.

If you want to merge it or not we are ok with it. Keep up the great work.

jsumners commented:
> This issue only happens when ldapjs is bundle with a global strict mode.

We don't support bundling. It's also a very bad practice to force transitive code (i.e. "not your code") into strict mode. The Node.js runtime does not do this for exactly the reason you have discovered: there is lots of code that will not work.

> I understand if you don't want to change existing code. We are using patch-package in our app to apply this fix.

Given that we don't support bundling, and the v2.x branch is soon to be dead, and that it is not possible to write a unit test to induce the issue, I think we will close this PR. Once v3 is released (hopefully by end of July? Don't hold me to it 🥺) I will close out a lot of issues, such as the one this PR attempts to address.

> Keep up the great work.

Thank you. And thank you for trying to resolve the problem with a PR.

jsumners commented:
⚠️ This issue has been locked due to age. If you have encountered a recent
problem that seems to be covered by this issue, please open a new issue.

Please include a [minimal reproducible example](https://en.wikipedia.org/wiki/Minimal_reproducible_example)
when opening a new issue.
